defi/single-token-amm-dex/src/lib.rs

```diff
@@ -0,0 +1,914 @@
+            }
+
+            // Burn metaBadge once all relative metaCandy has been claimed by end user 
+            // via "unstock_candy" function.
+            fn badge_burn(&mut self, meta_badge: Bucket) {
+                self.minter_badge.authorize(|auth| {meta_badge.burn_with_auth(auth)});
+            }
+
+            // Retrieve price of specific candy type.
+            fn candyprice(&mut self, candy_addr: Address) -> Decimal {
+                let price: Decimal;
+                
+                match self.candy_map.get(&candy_addr) {
+                    Some((_a,_b,c)) => price = *c,
+                    None => { 
+                        info!(" Candy not in stock! ");
+                        std::process::abort()
+                    }
+                };
+                
+                price
+            }
+
+            // Put candy to sell in vault whenever end user specify an exact number of candy to sell 
+            // using swap functions.
+            fn candyput_pri(
+                &mut self, 
+                prc_in: Decimal, 
+                prc_out: Decimal, 
+                addr_in: Address, 
+                candy: Bucket
+            ) -> (Decimal,Decimal) {
+                let candy_addr = candy.resource_address();
+                let mut candy_amnt = candy.amount();   
+                           
+                let nmbr = candy.amount()*prc_in/prc_out;
+                
+                if prc_in == dec!(0) { 
+                    candy_amnt = CandyDex::adjust_fee(self, candy_amnt); 
+                }
+                
+                let amount = CandyDex::candy_sum(self, candy_amnt, candy_addr, addr_in, dec!(1));
+
+                let v = self.candy_vaults.get_mut(&candy_addr).unwrap();
+
+                v.put(candy);
+
+                (nmbr,amount)
+            }
+
+            // Put candy to sell in vault whenever the number of candy to sell is determined by 
+            // protocol.
+            fn candyput_sec(
+                &mut self, 
+                amnt: Decimal, 
+                addr: Address, 
+                prc_in: Decimal, 
+                prc_out: Decimal, 
+                candy: Bucket
+            ) -> Bucket {
+                let candy_addr = candy.resource_address();
+                let amount = candy.amount();                
+                let amnt_in: Decimal;
+                
+                if prc_in == dec!(0) {
+                    amnt_in = CandyDex::candy_sum(self, amnt, addr, candy_addr, dec!(1));
+                } else {
+                    amnt_in = amnt*prc_in/prc_out;
+                }           
+                assert!( amnt_in <= amount, "Not enough input amount ");
+
+                let v = self.candy_vaults.get_mut(&candy_addr).unwrap();
+
+                v.put(candy);
+
+                let candy_output: Bucket = v.take(amount-amnt_in);
+
+                candy_output
```

Comment:
It would be simpler like this: (don't forget to put `mut` before `candy: Bucket`)
```suggestion
                v.put(candy.take(amnt_in));
                candy
```